### PR TITLE
Feature 1180 GenEnsProd wrapper

### LINKED ISSUE #1180: New Wrapper: GenEnsProd
## Describe the New Feature ##
dtcenter/MET#1904 adds an initial version of the new Gen-Ens-Prod tool to the met-10.1.0-beta3 MET development release. This issue adds a wrapper for calling that Gen-Ens-Prod tool. Just like Gen-Ens-Prod is a subset of the Ensemble-Stat tool, it's wrapper can largely be a subset of the existing Ensemble-Stat wrapper.

From @georgemccabe:
The new wrapper will be able to borrow existing components from other wrappers to implement all of the functionality. The wrapper will support the following functionality:

- Obtain a list of input files, write a file list file, and pass to -ens argument
- Set environment variables to override values in MET config file (all dictionaries and variables in the new config file are supported in other wrappers)
- Pass wrapped config file as -config argument
- Build output file path from template/dir and pass to -out argument
- Set optional -ctrl path based on template/dir (GEN_ENS_PROD_CTRL_[TEMPLATE/DIR] or GEN_ENS_PROD_CONTROL_[TEMPLATE/DIR])

### Acceptance Testing ###
*List input data types and sources.*
*Describe tests required for new functionality.*

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
2793541

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: @georgemccabe 
- [x] Select **scientist(s)** or **no scientist** required: @j-opatz 

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Select **Repository** and/or **Organization** level **Project(s)** or add **alert: NEED PROJECT ASSIGNMENT** label
- [x] Select **Milestone** as the next official version or **Future Versions**

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- dtcenter/MET#1904 creates the Gen-Ens-Prod tool.
- No other downstream issues needed. Although training materials should be developed eventually.

## New Feature Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [ ] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [ ] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [ ] Add/update log messages for easier debugging.
- [ ] Add/update unit tests.
- [ ] Add/update documentation.
- [ ] Push local changes to GitHub.
- [ ] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [ ] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)** and **Linked issues**
Select: **Repository** level development cycle **Project** for the next official release
Select: **Milestone** as the next official version
- [ ] Iterate until the reviewer(s) accept and merge your changes.
- [ ] Delete your fork or branch.
- [ ] Close this issue.

### PR DESCRIPTION
**Pull request by georgemccabe**
### Code Changes
* Added new wrapper GenEnsProd and met_tool_wrapper example use case
* Enhanced logic for handling wrapped MET config files
  * handle_met_config_dict function call can replace complex functions (i.e. handle_nmep_smooth_dict in ensemble_stat_wrapper.py)
  * added support for supplying "dictlist" (list of dictionaries - currently only 1 list item is supported)
* Added some support for getting full list of field levels instead of a single field item for each name/level (needed for future work)
* Added parent class LoopTimesWrapper to simplify logic common to many wrappers (used for new wrapper - future work to apply to other wrappers)
* Rearranged RuntimeFreqWrapper logic to check for errors in runtime frequency settings to allow default value for wrappers to be set (related to future work mentioned above)
* Changed MTD wrapper to inherit from CompareGriddedWrapper instead of MODEWrapper because it no longer needs anything from that wrapper (simplifies logic in MTD)
* Changed default value for argument in all wrappers' init function from immutable value to prevent pep8 complaints

### Testing
* Added empty files for use in unit tests (under internal_tests/data)
* Added missing tests for EnsembleStat wrapper

### Documentation
* Moved some common info from new use case documentation file to System Configuration section
* Minor updates to contributor's guide for creating a new wrapper
* Minor fixing to formatting in documentation (bold text not rendered properly)

## Pull Request Testing ##

- [X] Describe testing already performed for these changes:</br>

Ran met_tool_wrapper/GenEnsProd/GenEnsProd.conf use case many times, changing different configuration variables and verifying that the correct result occurred from each run.

- [X] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

I apologize for the numerous changes in this PR. Many are minor changes or adding empty files. The code changes could be skimmed for obvious errors, but the main focus of the review should be on the new wrapper and running that use case.

* Try to run the use case on seneca and play with changing settings to see if the results are as expected (consider recruiting a scientist to also test this portion)
* Review the documentation changes
* Ensure all automated tests pass (see below for expected diffs)

- [X] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

- [X] Do these changes include sufficient testing updates? **[Yes]**

- [X] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

The new data generated from the new use case should created new output that is not in the truth data. The only failed test should be the pull request run for the difference test for the new use case.

- [X] Please complete this pull request review by **10/28/2021**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [X] Complete the PR definition above.
- [X] Ensure the PR title matches the feature or bugfix branch name.
- [X] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Linked issues** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

**Comment by JohnHalleyGotway** [comment]
@georgemccabe I started looking at the PR today and have a question related to these config file options that show up in MET's GenEnsProd config file.
```
//
// May be set separately in each "field" entry
//
censor_thresh = [];
censor_val    = [];
cat_thresh    = [];
nc_var_str    = "";
```
MET parses these separately for each "ens.field" array entry. They are defined with default empty values at the top config file context level so that they are actually defined. If the user overrides the default setting in one of the field entries then that takes precedence.

I see corresponding METplus config entries: GEN_ENS_PROD_CENSOR_THRESH, GEN_ENS_PROD_CENSOR_VAL, GEN_ENS_PROD_CAT_THRESH, GEN_ENS_PROD_NC_VAR

Defining these at the top level of config file context does work if only processing a single variable. But if processing multiple variables, they'd need to be defined inside each "ens.field" array entry.

I just wanted to clarify the functionality for these types of config options. If processing multiple variables, would they use the "ENS_VAR{N}_OPTIONS" to specify the "cat_thresh" for example?

**Comment by georgemccabe** [comment]
@JohnHalleyGotway, if 2 fields are specified and those values are set on the top level, would they apply to each field or would they still need to be specified for each field separately? If the latter is true, then I think we should get rid of the 4 METplus variables listed and force the user to define this info with the field info using ENS_VAR<n>_OPTIONS as described here: https://metplus.readthedocs.io/en/latest/Users_Guide/systemconfiguration.html#field-info 

@CPKalb , @DanielAdriaansen , @j-opatz any thoughts?

File: docs/Contributors_Guide/basic_components.rst [hunk 1]
```diff
@@ -1,23 +1,50 @@
...
-            self.app_name = 'example'
-            self.app_path = os.path.join(self.config.getdir('MET_BIN_DIR'),
-                                                             self.app_name)
-
-The above code block is an excerpt from the ExampleWrapper, found in ush/example_wrapper.py. The class name should always be the item that is passed into the METplus configuration variable list PROCESS_LIST with 'Wrapper' at the end. The text 'CommandBuilder' in parenthesis makes ExampleWrapper a subclass of CommandBuilder. In the __init__ function, the line starting with 'super()' calls the parent class __init__ function.
+CommandBuilder is the parent class of all METplus wrappers.
+Every wrapper is a subclass of CommandBuilder or
+another subclass of CommandBuilder.
+For example, GridStatWrapper, PointStatWrapper, EnsembleStatWrapper,
+and MODEWrapper are all a subclass of CompareGriddedWrapper.
```
Comment (JohnHalleyGotway) [inline docs/Contributors_Guide/basic_components.rst]:
```suggestion
and MODEWrapper are all subclasses of CompareGriddedWrapper.
```

**Review by JohnHalleyGotway — APPROVED** [review]
I approve of this PR but made one minor suggested edit to wording.

I checked the tests and see that the only flagged error is the addition of the new test for this new wrapper:
- met_tool_wrapper:58 creates 3 new output files from GenEnsProd

I reviewed the 69 modified files and see:
- New met_tool_wrapper test addition.
- Contributor's Guide updates to make it easier to add new wrappers.
- User's Guide updates for the GenEnsProd wrapper.
- The new gen_ens_prod_wrapper.py wrapper itself.
- Several new GRIB input files in internal_tests - just empty files for testing command line generation.
- Changes to the ensemble-stat wrapper to move common config parsing functionality into one spot.
- Minor changes to many wrappers setting config_overrides from '{}' to 'None'.
- Addition of the loop_times_wrapper.py.
- Restructuring of class hierarchy to use CompareGriddedWrapper.

**Pull request merged**